### ISSUE TITLE
Preparing Polygon Changes Contains Results

### ISSUE THREAD

**Issue by bixb0012** [Toblerity/Shapely#268]
Start with a simple polygon and call contains with an empty polygon:
```python
>>> from shapely import wkt
>>> from shapely import prepared
>>> polygon = wkt.loads('POLYGON ((0 0, 0 1, 1 1, 1 0, 0 0))')
>>> polygon_empty = wkt.loads('POLYGON EMPTY')
>>> polygon.contains(polygon_empty)
False
```
So far, so good.  The issue I am running into is the result after preparing the simple polygon:
```python
>>> polygon_prepared = prepared.prep(polygon)
>>> polygon_prepared.contains(polygon_empty)
True
```
Expected behavior?  Bug?

Thanks.

**Comment by bixb0012** [comment]
Forgot to include:
```python
>>> sys.version
'2.7.8 (default, Jun 30 2014, 16:03:49) [MSC v.1500 32 bit (Intel)]'
>>> shapely.__version__
'1.5.6'
```

**Comment by snorfalorpagus** [comment]
I can reproduce this peculiarity using your code. However, the equivalent C seems to behave as expected (returning "false" for both). A bug in Shapely?

```
#include <stdio.h>

#include "geos_c.h"

int main() {
    GEOSWKTReader *reader;
    GEOSGeometry* polygon1;
    GEOSGeometry* polygon2;
    const GEOSPreparedGeometry* prepared1;
    
    initGEOS(NULL, NULL);
    
    reader = GEOSWKTReader_create();
    
    polygon1 = GEOSWKTReader_read(reader, "POLYGON ((0 0, 0 1, 1 1, 1 0, 0 0))");
    polygon2 = GEOSWKTReader_read(reader, "POLYGON EMPTY");
    
    char a = GEOSContains(polygon1, polygon2);
    if (a == 0) {
        printf("false\n");
    } else if(a == 1) {
        printf("true\n");
    } else {
        printf("error\n");
    }
    
    prepared1 = GEOSPrepare(polygon1);
    
    char b = GEOSPreparedContains(prepared1, polygon2);
    if (b == 0) {
        printf("false\n");
    } else if(b == 1) {
        printf("true\n");
    } else {
        printf("error\n");
    }
    
    return 0;
}
```

**Comment by snorfalorpagus** [comment]
This may be a bug with the Windows build of `libgeos`. On OS X (homebrew) I get:

```
Python 3.4.3 (default, Apr  7 2015, 08:05:21) 
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from shapely import wkt
>>> from shapely import prepared
>>> polygon = wkt.loads('POLYGON ((0 0, 0 1, 1 1, 1 0, 0 0))')
>>> polygon_empty = wkt.loads('POLYGON EMPTY')
>>> polygon.contains(polygon_empty)
False
>>> polygon_prepared = prepared.prep(polygon)
>>> polygon_prepared.contains(polygon_empty)
False
```

I realise when I tried the C example above, I didn't link against the same library that shapely uses. Are you using the Windows build from http://www.lfd.uci.edu/~gohlke/pythonlibs/ ?

**Comment by bixb0012** [comment]
Yes.

I just did a clean 64-bit Conda install with [SciTools'](http://scitools.org.uk) Shapely 1.5.8dev0 package using geos-3.4.2-3, and it returns True as well on the prepared polygon.
 ```python
>>> sys.version
'2.7.9 |Continuum Analytics, Inc.| (default, Dec 18 2014, 16:57:52) [MSC v.1500 64 bit (AMD64)]'
>>> shapely.__version__
'1.5.8'
```

**Comment by bixb0012** [comment]
I am quickly coming around to the idea it is a bug with the Windows build of `libgeos`.  Looking at how other prepared geometry operators interact with an empty geometry, there seems to be a larger issue between prepared geometries and empty geometries.
```python
>>> from shapely import wkt
>>> from shapely import prepared
>>> polygon = wkt.loads('POLYGON((0 0, 1 0, 1 1, 0 1, 0 0))')
>>> polygon_empty = wkt.loads('POLYGON EMPTY')
>>> polygon.intersects(polygon_empty)
False
>>> polygon.contains(polygon_empty)
False
>>> polygon_prepared = prepared.prep(polygon)
>>> polygon_prepared.intersects(polygon_empty)
True
>>> polygon_prepared.contains(polygon_empty)
True
>>> polygon_prepared.contains_properly(polygon_empty)
True
```
I am not sure where to take it from here, but it doesn't seem to be a Shapely issue, so I will close this out.

**Comment by bixb0012** [comment]
Just tried similar test using GeoDjango pointing to the same geos_c.dll on my Windows machine:
```python
>>> from django.contrib.gis.geos import GEOSGeometry
>>> pg = GEOSGeometry('POLYGON((0 0, 1 0, 1 1, 0 1, 0 0))')
>>> pg_e = GEOSGeometry('POLYGON EMPTY')
>>> pg.ewkt
u'POLYGON ((0.0000000000000000 0.0000000000000000, 1.0000000000000000 0.0000000000000000, 1.0000000000000000 1.0000000000000000, 0.0000000000000000 1.0000000000000000, 0.0000000000000000 0.0000000000000000))'
>>> pg_e.ewkt
u'POLYGON EMPTY'
>>> pg.contains(pg_e)
False
>>> pg_p = pg.prepared
>>> pg_p.contains(pg_e)
False
>>>
```
This does appear specific to Shapely and not the GEOS library being used.

**Comment by snorfalorpagus** [comment]
The return type of `GEOSPreparedContains` is `char` not `int`. I tried changing the restype to `c_char`.
https://github.com/libgeos/libgeos/blob/aa57155d18f8cbb33b5090c5ea4bae349abf4db2/capi/geos_c.cpp#L1042
https://github.com/Toblerity/Shapely/blob/master/shapely/ctypes_declarations.py#L309

I also modified the `BinaryPredicate` (`shapely/predicate.py`) class to print the return value:
```
class BinaryPredicate(Delegating):
    def __call__(self, this, other, *args):
        self._validate(this)
        self._validate(other, stop_prepared=True)
        try:
            ret = self.fn(this._geom, other._geom, *args)
            print('return', ret)
            return ret
        except PredicateError as err:
            # Dig deeper into causes of errors.
            self._check_topology(err, this, other)
```
Here's what I get:
```
('return', 0)
polygon contains empty polygon: False
('return', 1)
polygon contains POINT(0.5 0.5): True
('return', '\x00')
prepared polygon contains empty polygon: True
```
The `GEOSPreparedContains` function seems to be returning a NULL character, which is very strange. Not sure what is going on here, or why this should only happen on Windows. I tried compiling my C example against the same GEOS library that shapely is using and it functions correctly. I do think this is a shapely bug, but I'm not sure what's going on yet.

**Comment by sgillies** [comment]
Closed via #269. Thanks for the great bug report!

**Comment by bixb0012** [comment]
@sgillies, no problem, glad to help improve an already fine product.
